### ISSUE TITLE
Overall robustness and support of miniflare is severely lacking

### ISSUE THREAD

**Issue by dkfdkdodkms** [cloudflare/miniflare#579]
Everyday, I find one or more rather serious issues that need to be worked around. Here are two just from this morning:

> "dkfdkdodkms — Today at 11:06 AM
> testing locally using queues, i am getting message.ack is not a function. body, id, and timestamp properties are there. do i need to use experimental?"
> 
> "dkfdkdodkms — Today at 11:06 AM
> hm, so i tested with experimental local and various things break, so i am hoping the non-experimental version is more accurate. for example: when using the non-experimental version, i have to pipe uploads through a transform in order to put to r2 (probably due to a bug in miniflare), or i get an error. using the experimental version, if i pipe through the same transform, i get an error saying "readable stream must have a known length" (again, probably due to a bug in miniflare, as there is no apparent way of specifying a length manually). at this point, i don't know if i should use a transform or not for live, as both methods should probably be working (with and without a transform). this all appears to be extremely alpha (both experimental and non-experimental version), and i'm a bit surprised by that. i'm further surprised that miniflare wasn't even started by cloudflare, as one would think the need for local development would have been understood from the onset. anyway, i request cloudflare invest more resources into miniflare, as the current robustness is about what i would expect from an individual's hobby project (perhaps even worse)."

Unfortunately, this issue doesn't appear to be limited to miniflare. Shortly after my initial post a user posted about the new TCP sockets feature, and this is what followed:

> dkfdkdodkms — Today at 11:07 AM
> nice, but i would rather them invest in fixing what's broken
> AA — Today at 11:07 AM
> except i needed that exact functionality 😄
> dkfdkdodkms — Today at 11:08 AM
> until you find it's broken too 😉
> AA — Today at 11:08 AM
> theres been a few misses, indeed
> 
> then...
> 
> zizi started a thread: websocket client close casue worker throw error — Today at 11:13 AM

So, my suggestion to Cloudflare in resolving their quite significant code quality issue is this: 

**Please stop. Take a breath and fix what you have. Then, and only then, add new features. Repeat.**

**Comment by penalosa** [comment]
Thanks for bringing these issues up in Discord! If you can, it's also helpful to raise issues on GitHub—that's the best way to flag specific bugs, since it's easy to miss things in Discord. I'm going to close this issue for now, but rest assured that we do take Miniflare (and Wrangler) seriously, and they do have significant resources invested in them. To explain a bit more about the `--local`/`--experimental-local` difference:
* `wrangler dev --local` uses `miniflare@2`, which is a NodeJS-based emulator for the Cloudflare Workers runtime environment.
* `wrangler dev --experimental-local` uses `miniflare@3`, the new version of Miniflare based on the open source Cloudflare Workers runtime (`workerd`). This generally maps more closely to the behaviour of deployed workers.
I hope this is helpful! Let me know if you have any more questions, and I'd be happy to answer them.

**Comment by johnbrandle** [comment]
@DaniFoldi Thanks for your response. I appreciate that this repository is open-source, and that anyone can contribute to it, but it shouldn't be a necessity, or expected. After all, this exists to support Cloudflare products; Cloudflare should be at the forefront of it's development, if only for self-interest. Essentially, this is the first time in 20 years that I have had a need to examine the code of a company's offerings due to buggy and/or unexpected behavior. That's just not normal. Especially, not for a company of Cloudflare's size. Small startups and individuals have less buggy code...

If I had to guess, I would imagine this is likely due to the speed at which Cloudflare is adding and/or changing features, making keeping code such as this in a robust state very challenging.

I would also say that unless the product is in alpha, things like documented methods not existing on 'message' should not be a thing. Perhaps the bar needs to be raised for what is considered beta. Basically, it looks like more internal testing is needed. Is any actually occurring? Serious question.

Lastly, there are a lot of bugs (most unreported outside of discord), and I consider all of those buckets to be actual bugs. Those are all things which will cause serious issues with local development, AKA, serious development. And, cause serious headache when deploying due to behavioral differences. The way things are now, it basically just attracts those looking to do trivial things and/or those who are primarily interested in the free services aspect...not exactly the hard to appease types.

@penalosa Did you know it's been two weeks since a bug has been resolved on here? Well, aside from this one, which was resolved rather quickly... Also, I have switched over to --experimental-local a couple times now, and every time I do everything breaks in several different ways...it's frightening how different these two modes are. I hesitate to move to --experimental-local given some of the issues I've encountered, not to mention the fact that it's still experimental.

Basically, Cloudflare needs to level up its development. Test, test, test... and maybe actually fix a thing or two as if it was a priority.

**Comment by dkfdkdodkms** [comment]
@DaniFoldi Thanks for your response. I appreciate that this repository is open-source, and that anyone can contribute to it, but it shouldn't be a necessity, or expected. After all, this exists to support Cloudflare products; Cloudflare should be at the forefront of it's development, if only for self-interest. Essentially, this is the first time in 20 years that I have had a need to examine the code of a company's offerings due to buggy and/or unexpected behavior. That's just not normal. Especially, not for a company of Cloudflare's size. Small startups and individuals have less buggy code...

If I had to guess, I would imagine this is likely due to the speed at which Cloudflare is adding and/or changing features, making keeping code in a robust state very challenging.

I would also say that unless the product is in alpha, things like documented methods not existing on 'message' should not be a thing. Perhaps the bar needs to be raised for what is considered beta. Basically, it looks like more internal testing is needed. Is any actually occurring? Serious question.

Lastly, there are a lot of bugs (most unreported outside of discord), and I consider all of those buckets to be actual bugs. Those are all things which will cause serious issues with local development, AKA, serious development. And, cause serious headache when deploying due to behavioral differences. The way things are now, it basically just attracts those looking to do trivial things and/or those who are primarily interested in the free services aspect...not exactly the hard to appease types.

@penalosa Did you know it's been two weeks since a bug has been resolved on here? Well, aside from this one, which was resolved rather quickly... Also, I have switched over to --experimental-local a couple times now, and every time I do everything breaks in several different ways...it's frightening how different these two modes are. I hesitate to move to --experimental-local given some of the issues I've encountered, not to mention the fact that it's still experimental.

Basically, Cloudflare needs to level up its development. Test, test, test... and maybe actually fix a thing or two as if it was a priority.
